pkg/fatfs/fatfs_vfs/fatfs_vfs.c

```diff
@@ -0,0 +1,443 @@
+/*
+ * Copyright (C) 2017 HAW-Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     fs
+ * @{
+ *
+ * @file
+ * @brief       FatFs wrapper for vfs
+ *
+ * @author      Michel Rottleuthner <michel.rottleuthner@haw-hamburg.de>
+ *
+ * @}
+ */
+
+
+#include <fcntl.h>
+#define __LINUX_ERRNO_EXTENSIONS__
+#include <errno.h>
+#include <inttypes.h>
+#include <sys/stat.h> /* for struct stat */
+#include <string.h>
+
+#include "fs/fatfs.h"
+
+#include "kernel_defines.h"
+#include "time.h"
+
+#define ENABLE_DEBUG (0)
+#include <debug.h>
+
+static int fatfs_err_to_errno(int32_t err);
+static void _fatfs_time_to_timespec(WORD fdate, WORD ftime, time_t *time);
+
+static int _mount(vfs_mount_t *mountp)
+{
+    /* if one of the lines below fail to compile you probably need to adjust
+       vfs buffer sizes ;) */
+    BUILD_BUG_ON(VFS_DIR_BUFFER_SIZE < sizeof(DIR));
+    BUILD_BUG_ON(VFS_FILE_BUFFER_SIZE < sizeof(fatfs_file_desc_t));
+
+    fatfs_desc_t *fs_desc = (fatfs_desc_t *)mountp->private_data;
+
+    char volume_str[FATFS_MAX_VOL_STR_LEN];
+    sprintf(volume_str, "%d:/", fs_desc->vol_idx);
+
+    DEBUG("mounting file system of volume '%s'\n", volume_str);
+    FRESULT res = f_mount(&fs_desc->fat_fs, volume_str, 1);
+
+    if (res == FR_OK) {
+        DEBUG("[OK]");
+    }
+    else {
+        DEBUG("[ERROR]");
+    }
+
+    return fatfs_err_to_errno(res);
+}
+
+static int _umount(vfs_mount_t *mountp)
+{
+    fatfs_desc_t *fs_desc = mountp->private_data;
+
+    DEBUG("fatfs_vfs.c: _umount: private_data = %p\n", mountp->private_data);
+
+    char volume_str[FATFS_MAX_VOL_STR_LEN];
+    sprintf(volume_str, "%d:/", fs_desc->vol_idx);
+
+    DEBUG("unmounting file system of volume '%s'\n", volume_str);
+    FRESULT res = f_mount(NULL, volume_str, 1);
+
+    if (res == FR_OK) {
+        DEBUG("[OK]");
+        memset(&fs_desc->fat_fs, 0, sizeof(fs_desc->fat_fs));
+    }
+    else {
+        DEBUG("[ERROR]");
+    }
+
+    return fatfs_err_to_errno(res);
+}
+
+static int _unlink(vfs_mount_t *mountp, const char *name)
+{
+    (void) mountp;
+    return fatfs_err_to_errno(f_unlink(name));
+}
+
+static int _rename(vfs_mount_t *mountp, const char *from_path,
+                   const char *to_path)
+{
+    (void) mountp;
+    return fatfs_err_to_errno(f_rename(from_path, to_path));
+}
+
+static int _open(vfs_file_t *filp, const char *name, int flags, mode_t mode,
+                 const char *abs_path)
+{
+    fatfs_file_desc_t *fd = (fatfs_file_desc_t *)&filp->private_data.buffer[0];
+    char fatfs_abs_path[FATFS_MAX_VOL_STR_LEN + VFS_NAME_MAX + 1];
```

Comment:
right - how about adding something like a `char path_str_working_buffer[FATFS_MAX_VOL_STR_LEN + VFS_NAME_MAX + 1]` to [fatfs_desc_t](https://github.com/MichelRottleuthner/RIOT/blob/fatfs_vfs/sys/include/fs/fatfs.h#L74)?